include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+      base_type::operator=(std::move(other));
+      return *this;
+    }
+  };
+
+  class DynamicWorkspace final : public Workspace<std::vector<alloc_type*>> {
+   public:
+    using base_type = Workspace<std::vector<alloc_type*>>;
+
+    inline DynamicWorkspace() : base_type() {}
+
+    inline DynamicWorkspace(pool_type* pool, size_t n, cudaStream_t ctx)
+        : base_type(pool, ctx) {
+      auto& buffers = this->buffers_;
+      buffers.resize(n);
+      pool->get_raw(buffers.begin(), buffers.end(), ctx);
+    }
+
+    DynamicWorkspace(const DynamicWorkspace&) = delete;
+    DynamicWorkspace& operator=(const DynamicWorkspace&) = delete;
+
+    inline DynamicWorkspace(DynamicWorkspace&& other)
+        : base_type(std::move(other)) {}
+    inline DynamicWorkspace& operator=(DynamicWorkspace&& other) {
+      base_type::operator=(std::move(other));
+      return *this;
+    }
+  };
+
+  const MemoryPoolOptions options;
+
+  MemoryPool(const MemoryPoolOptions& options) : options{options} {
+    MERLIN_CHECK(options.buffer_size > 0,
+                 "[HierarchicalKV] Memory pool buffer size invalid!");
+    MERLIN_CHECK(
+        options.base_stock <= options.max_stock,
+        "[hierarchicalKV] Initial reserve cannot exceed maximum reserve!");
+
+    // Create initial buffer stock.
+    stock_.reserve(options.max_stock);
+    while (stock_.size() < options.base_stock) {
+      stock_.emplace_back(Allocator::alloc(options.buffer_size));
+    }
+
+    // Create enough events, so we have one per potentially pending buffer.
+    ready_events_.resize(options.max_pending);
+    for (auto& ready_event : ready_events_) {
+      CUDA_CHECK(cudaEventCreate(&ready_event));
+    }
+
+    // Preallocate pending.
+    pending_.reserve(options.max_pending);
+  }
+
+  ~MemoryPool() {
+    // Make sure all queued tasks are complete.
+    await_pending();
+
+    // Free event and buffer memory.
+    for (auto& ready_event : ready_events_) {
+      CUDA_CHECK(cudaEventDestroy(ready_event));
+    }
+
+    for (auto& ptr : stock_) {
+      Allocator::free(ptr);
+    }
+  }
+
+  inline size_t buffer_size() const { return options.buffer_size; }
+
+  inline size_t max_batch_size(size_t max_item_size) const {
+    return options.buffer_size / max_item_size;
+  }
+
+  template <class T>
+  inline size_t max_batch_size() const {
+    return max_batch_size(sizeof(T));
+  }
+
+  size_t current_stock() const {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return stock_.size();
+  }
+
+  size_t num_pending() const {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return pending_.size();
+  }
+
+  void deplete(cudaStream_t ctx = nullptr) {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Collect any pending buffers first.
+    collect_pending_unsafe(ctx);
+
+    // Deplete remaining buffers.
+    for (auto& ptr : stock_) {
+      Allocator::free(ptr, ctx);
+    }
+    stock_.clear();
+  }
+
+  void replenish(cudaStream_t ctx = nullptr) {
```

Comment:
It's not recommended to keep any used API.

Reply:
This API, I need for testing.